Add Check Support for Java 21 Record Pattern : ParenPad

child of #14890

I have read check documentation: https://checkstyle.org/checks/whitespace/parenpad.html#ParenPad
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
PS D:\CS\test> javac src/RecordPatterns.java 
PS D:\CS\test> cat src/RecordPatterns.java
import java.util.List;

public class RecordPatterns {
    record Box<T>(T t) { }

    void Example1(Box<Box<String>> bbs) {
        switch (bbs) {
             case Box(  Box(  var s)) -> System.out.println(s); // expected violation
        }
        if (bbs instanceof Box( Box( var s))) { // violation
            System.out.println(s);
        }
    }
}
PS D:\CS\test> cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
 <module name="TreeWalker">
    <module name="ParenPad"/>
 </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.14.2-all.jar -c config.xml  src/RecordPatterns.java
Starting audit...
[ERROR] D:\CS\test\src\RecordPatterns.java:10:31: '(' is followed by whitespace. [ParenPad]
[ERROR] D:\CS\test\src\RecordPatterns.java:10:36: '(' is followed by whitespace. [ParenPad]
Audit done.
Checkstyle ends with 2 errors.
```

---

**Describe what you expect in detail.**
violations on  the record pattern used in case label

**Note:**
 violations on line 10 are due to https://github.com/checkstyle/checkstyle/issues/14747. the check violates record patterns under `EXPR` token even if the record pattern is not in the config. this is due to the deep scan
but still we should support this token in the check

---

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/ParenPadCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/ParenPadCheck.java
@@ -97,7 +97,9 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
  * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LAMBDA">
  * LAMBDA</a>,
  * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_DEF">
- * RECORD_DEF</a>.
+ * RECORD_DEF</a>,
+ * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_PATTERN_DEF">
+ * RECORD_PATTERN_DEF</a>.
  * </li>
  * </ul>
  * <p>
@@ -306,6 +308,7 @@ public class ParenPadCheck extends AbstractParenPadCheck {
             TokenTypes.SUPER_CTOR_CALL,
             TokenTypes.LAMBDA,
             TokenTypes.RECORD_DEF,
+            TokenTypes.RECORD_PATTERN_DEF,
         };
     }
 

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/whitespace/ParenPadCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/whitespace/ParenPadCheck.xml
@@ -26,7 +26,7 @@
                       type="com.puppycrawl.tools.checkstyle.checks.whitespace.PadOption">
                <description>Specify policy on how to pad parentheses.</description>
             </property>
-            <property default-value="ANNOTATION,ANNOTATION_FIELD_DEF,CTOR_CALL,CTOR_DEF,DOT,ENUM_CONSTANT_DEF,EXPR,LITERAL_CATCH,LITERAL_DO,LITERAL_FOR,LITERAL_IF,LITERAL_NEW,LITERAL_SWITCH,LITERAL_SYNCHRONIZED,LITERAL_WHILE,METHOD_CALL,METHOD_DEF,QUESTION,RESOURCE_SPECIFICATION,SUPER_CTOR_CALL,LAMBDA,RECORD_DEF"
+            <property default-value="ANNOTATION,ANNOTATION_FIELD_DEF,CTOR_CALL,CTOR_DEF,DOT,ENUM_CONSTANT_DEF,EXPR,LITERAL_CATCH,LITERAL_DO,LITERAL_FOR,LITERAL_IF,LITERAL_NEW,LITERAL_SWITCH,LITERAL_SYNCHRONIZED,LITERAL_WHILE,METHOD_CALL,METHOD_DEF,QUESTION,RESOURCE_SPECIFICATION,SUPER_CTOR_CALL,LAMBDA,RECORD_DEF,RECORD_PATTERN_DEF"
                       name="tokens"
                       type="java.lang.String[]"
                       validation-type="tokenSet">

--- a/src/main/resources/google_checks.xml
+++ b/src/main/resources/google_checks.xml
@@ -294,7 +294,7 @@
                     EXPR, LITERAL_CATCH, LITERAL_DO, LITERAL_FOR, LITERAL_IF, LITERAL_NEW,
                     LITERAL_SWITCH, LITERAL_SYNCHRONIZED, LITERAL_WHILE, METHOD_CALL,
                     METHOD_DEF, QUESTION, RESOURCE_SPECIFICATION, SUPER_CTOR_CALL, LAMBDA,
-                    RECORD_DEF"/>
+                    RECORD_DEF, RECORD_PATTERN_DEF"/>
     </module>
     <module name="OperatorWrap">
       <property name="option" value="NL"/>

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/ParenPadCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/ParenPadCheckTest.java
@@ -580,4 +580,58 @@ public class ParenPadCheckTest
                 getNonCompilablePath("InputParenPadCheckWhenExpression.java"), expected);
     }
 
+    @Test
+    public void testParenPadForRecordPattern() throws Exception {
+        final String[] expected = {
+            "14:40: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "16:40: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "16:60: " + getCheckMessage(MSG_WS_PRECEDED, ")"),
+            "22:40: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "22:47: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "26:46: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "26:61: " + getCheckMessage(MSG_WS_PRECEDED, ")"),
+            "26:73: " + getCheckMessage(MSG_WS_PRECEDED, ")"),
+            "31:40: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "31:47: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "31:62: " + getCheckMessage(MSG_WS_PRECEDED, ")"),
+            "31:74: " + getCheckMessage(MSG_WS_PRECEDED, ")"),
+            "42:23: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "44:23: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "44:38: " + getCheckMessage(MSG_WS_PRECEDED, ")"),
+            "52:30: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "52:37: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "56:36: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "56:51: " + getCheckMessage(MSG_WS_PRECEDED, ")"),
+            "56:63: " + getCheckMessage(MSG_WS_PRECEDED, ")"),
+            "61:30: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "61:37: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "61:52: " + getCheckMessage(MSG_WS_PRECEDED, ")"),
+            "61:64: " + getCheckMessage(MSG_WS_PRECEDED, ")"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputParenPadForRecordPattern.java"), expected);
+    }
+
+    @Test
+    public void testParenPadForRecordPatternWithSpaceOption() throws Exception {
+        final String[] expected = {
+            "14:40: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "("),
+            "14:58: " + getCheckMessage(MSG_WS_NOT_PRECEDED, ")"),
+            "18:59: " + getCheckMessage(MSG_WS_NOT_PRECEDED, ")"),
+            "23:40: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "("),
+            "23:46: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "("),
+            "23:59: " + getCheckMessage(MSG_WS_NOT_PRECEDED, ")"),
+            "23:70: " + getCheckMessage(MSG_WS_NOT_PRECEDED, ")"),
+            "29:61: " + getCheckMessage(MSG_WS_NOT_PRECEDED, ")"),
+            "29:72: " + getCheckMessage(MSG_WS_NOT_PRECEDED, ")"),
+            "33:40: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "("),
+            "41:30: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "("),
+            "41:48: " + getCheckMessage(MSG_WS_NOT_PRECEDED, ")"),
+            "45:37: " + getCheckMessage(MSG_WS_NOT_PRECEDED, ")"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputParenPadForRecordPatternWithSpaceOption.java"),
+                expected);
+    }
+
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadForRecordPattern.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadForRecordPattern.java
@@ -1,0 +1,73 @@
+/*
+ParenPad
+option = (default)nospace
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.whitespace.parenpad;
+
+public class InputParenPadForRecordPattern {
+
+    void test(Object obj) {
+        if (obj instanceof ColoredPoint(Point p, String c)) {}
+        if (obj instanceof ColoredPoint( Point p, String c)) {}
+        // violation above, ''(' is followed by whitespace'
+        if (obj instanceof ColoredPoint( Point p, String c )) {}
+        // 2 violations above:
+        //             ''(' is followed by whitespace'
+        //             '')' is preceded with whitespace'
+
+        if (obj instanceof ColoredPoint(Point(int x, int _), String c)) {}
+        if (obj instanceof ColoredPoint( Point( int x, int _), String c)) {}
+        // 2 violations above:
+        //             ''(' is followed by whitespace'
+        //             ''(' is followed by whitespace'
+        if (obj instanceof ColoredPoint(Point( int x, int _ ), String c )) {}
+        // 3 violations above:
+        //             ''(' is followed by whitespace'
+        //             '')' is preceded with whitespace'
+        //             '')' is preceded with whitespace'
+        if (obj instanceof ColoredPoint( Point( int x, int _ ), String c )) {}
+        // 4 violations above:
+        //             ''(' is followed by whitespace'
+        //             ''(' is followed by whitespace'
+        //             '')' is preceded with whitespace'
+        //             '')' is preceded with whitespace'
+
+    }
+    void test2(Object obj) {
+        switch (obj) {
+            case ColoredPoint(Point p, String c) -> {}
+            case Point( int x, int y) when x == 0 -> {}
+            // violation above, ''(' is followed by whitespace'
+            case Point( int x, int y ) -> {}
+            // 2 violations above:
+            //             ''(' is followed by whitespace'
+            //             '')' is preceded with whitespace'
+            default -> {}
+        }
+        switch (obj) {
+            case ColoredPoint(Point(int x, int _), String c) when x == 0-> {}
+            case ColoredPoint( Point( int x, int _), String c) when c.isEmpty() -> {}
+            // 2 violations above:
+            //             ''(' is followed by whitespace'
+            //             ''(' is followed by whitespace'
+            case ColoredPoint(Point( int x, int _ ), String c ) when x == 0 -> {}
+            // 3 violations above:
+            //             ''(' is followed by whitespace'
+            //             '')' is preceded with whitespace'
+            //             '')' is preceded with whitespace'
+            case ColoredPoint( Point( int x, int _ ), String c ) when c.isEmpty() -> {}
+            // 4 violations above:
+            //             ''(' is followed by whitespace'
+            //             ''(' is followed by whitespace'
+            //             '')' is preceded with whitespace'
+            //             '')' is preceded with whitespace'
+            default -> {}
+        }
+    }
+
+    record ColoredPoint(Point p, String color) {}
+    record Point(int x, int y) {}
+}

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadForRecordPatternWithSpaceOption.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadForRecordPatternWithSpaceOption.java
@@ -1,0 +1,55 @@
+/*
+ParenPad
+option = space
+tokens = RECORD_PATTERN_DEF
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.whitespace.parenpad;
+
+public class InputParenPadForRecordPatternWithSpaceOption {
+
+    void test(Object obj) {
+        if (obj instanceof ColoredPoint(Point p, String c)) {}
+        // 2 violations above:
+        //             ''(' is not followed by whitespace'
+        //             '')' is not preceded with whitespace'
+        if (obj instanceof ColoredPoint( Point p, String c)) {}
+        // violation above, '')' is not preceded with whitespace'
+        if (obj instanceof ColoredPoint( Point p, String c )) {}
+
+
+        if (obj instanceof ColoredPoint(Point(int x, int _), String c)) {}
+        // 4 violations above:
+        //             ''(' is not followed by whitespace'
+        //             ''(' is not followed by whitespace'
+        //             '')' is not preceded with whitespace'
+        //             '')' is not preceded with whitespace'
+        if (obj instanceof ColoredPoint( Point( int x, int _), String c)) {}
+        // 2 violations above:
+        //             '')' is not preceded with whitespace'
+        //             '')' is not preceded with whitespace'
+        if (obj instanceof ColoredPoint(Point( int x, int _ ), String c )) {}
+        // violation above, ''(' is not followed by whitespace'
+        if (obj instanceof ColoredPoint( Point( int x, int _ ), String c )) {}
+
+    }
+
+    void test2(Object obj) {
+        switch (obj) {
+            case ColoredPoint(Point p, String c) -> {}
+            // 2 violations above:
+            //             ''(' is not followed by whitespace'
+            //             '')' is not preceded with whitespace'
+            case Point( int x, int y) when x == 0 -> {}
+            // violation above, '')' is not preceded with whitespace'
+            case Point( int x, int y ) -> {}
+            default -> {}
+        }
+    }
+
+
+    record ColoredPoint(Point p, String color) {}
+    record Point(int x, int y) {}
+}

--- a/src/xdocs/checks/whitespace/parenpad.xml
+++ b/src/xdocs/checks/whitespace/parenpad.xml
@@ -92,6 +92,8 @@
                     LAMBDA</a>
                 , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_DEF">
                     RECORD_DEF</a>
+                , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_PATTERN_DEF">
+                    RECORD_PATTERN_DEF</a>
                   .
               </td>
               <td>
@@ -139,6 +141,8 @@
                     LAMBDA</a>
                 , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_DEF">
                     RECORD_DEF</a>
+                , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_PATTERN_DEF">
+                    RECORD_PATTERN_DEF</a>
                   .
               </td>
               <td>3.0</td>

